broadinstitute/pearl
# [JN-461] Designer validation

This adds very simple unenforced validation to a few areas in the survey question designer.

![Screenshot 2023-09-15 at 11 57 05 AM](https://github.com/broadinstitute/juniper/assets/7257391/b5b732e8-a071-4a97-a5fb-dc3c1964124d)

TO TEST:

Try adding a new question or editing an existing one. Required fields should have a red asterisk next to the field label and the input area should by styled with a warning if it's empty. Once some text is entered, the input area should return to normal styling.

_______

~~This is super rough, but this one has been slow to implement so I wanted to start getting feedback before I put more time into it. There aren't any tests yet, and I want to do a bit of refactoring for the `onChange` stuff which would allow me to push the overall validation check down into the `FormDesigner` component where it belongs. But for now, you get the idea.~~

~~The main challenge for implementing this was the designer tab doesn't have a concept of saving an individual question. You can freely click around the table of contents without actually committing to your changes. I think that's good from a UX perspective, because it reduces the number of clicks required of the user, but this means that we have to disable the use of the table of contents until the user has fixed any errors. Is this too restrictive/burdensome? Is the validation warning itself too loud?~~

~~Another issue is that the validation checks only occur when the `QuestionDesigner` is rendered for a question. This means that if you close the page out and load from a draft, any validation errors are lost (it doesn't re-validate the entire survey). I'm thinking of this type of validation as "best effort" rather than airtight, because none of these validation errors are actually going to cause problems when the survey is rendered. But it at least encourages decent field hygiene.~~

~~Final thought: I'm wondering if overall this is more trouble than it's worth right now. Even without this, we're still preventing all of the breaking cases (like invalid survey json, and requiring stable IDs for new questions)~~


~~https://github.com/broadinstitute/juniper/assets/7257391/5584c733-3389-446a-9e22-2cf4e41a9d8c~~
